packages/material/src/additional/MaterialListWithDetailRenderer.tsx

```diff
@@ -24,213 +24,115 @@
 */
 import {
   and,
-  ArrayLayoutProps,
   composePaths,
   computeLabel,
   createDefaultValue,
   findUISchema,
-  getData,
-  isPlainLabel,
   isObjectArray,
-  JsonFormsState,
-  JsonSchema,
-  mapDispatchToArrayControlProps,
-  mapStateToArrayLayoutProps,
+  isPlainLabel,
+  mapStateToJsonFormsRendererProps,
+  OwnPropsOfControl,
   RankedTester,
   rankWith,
-  Resolve,
   uiTypeIs
 } from '@jsonforms/core';
-import { ResolvedJsonForms } from '@jsonforms/react';
 import {
-  Avatar,
+  ctxDispatchToArrayControlProps,
+  ctxToArrayLayoutProps,
+  ResolvedJsonFormsDispatchRenderer,
+  useJsonForms
+} from '@jsonforms/react';
+import {
   Grid,
   Hidden,
-  IconButton,
   List,
-  ListItem,
-  ListItemAvatar,
-  ListItemSecondaryAction,
-  ListItemText,
   Typography
 } from '@material-ui/core';
-import DeleteIcon from '@material-ui/icons/Delete';
-import find from 'lodash/find';
 import map from 'lodash/map';
 import range from 'lodash/range';
-import React from 'react';
-import { connect } from 'react-redux';
+import React, { useState } from 'react';
 import { ArrayLayoutToolbar } from '../layouts/ArrayToolbar';
+import ListWithDetailMasterItem from './ListWithDetailMasterItem';
 
-interface MaterialListWithDetailRendererState {
-  selectedIndex: number;
-}
-export class MaterialListWithDetailRenderer extends React.Component<
-  ArrayLayoutProps,
-  MaterialListWithDetailRendererState
-> {
-  state: MaterialListWithDetailRendererState = {
-    selectedIndex: undefined
-  };
-  addItem = (path: string, value: any) => this.props.addItem(path, value);
-  removeItem = (path: string, value: number) => () => {
-    this.props.removeItems(path, [value])();
-    if (this.state.selectedIndex === value) {
-      this.state.selectedIndex = undefined;
-    } else if (this.state.selectedIndex > value) {
-      this.state.selectedIndex--;
+export const MaterialListWithDetailRenderer = (props: OwnPropsOfControl) => {
+  const [selectedIndex, setSelectedIndex] = useState(undefined);
+  const ctx = useJsonForms();
+  const { required, visible, errors, path, data, schema, label, uischema, uischemas } = ctxToArrayLayoutProps(ctx, props);
+  const { addItem, removeItems } = ctxDispatchToArrayControlProps(ctx.dispatch);
+  const handleRemoveItem = (p: string, value: any) => () => {
```

Comment:
If I understand the life cycle correctly, then the handleRemoveItem will be recreated every time we run through here and thus the ListWithDetailMasterItem will be re-rendered.

Reply:
Yes, that's the case currently, but we should be able to memoize it with `useCallback`, so that we can avoid any re-render